app/js/RequestLogger.js

```diff
@@ -0,0 +1,83 @@
+const logger = require('logger-sharelatex')
+const metrics = require('metrics-sharelatex')
+
+module.exports = {
+  errorHandler,
+  middleware
+}
+
+function errorHandler(err, req, res, next) {
+  req._logInfo.set('error', err)
+  res
+    .send(err.message)
+    .status(500)
+    .end()
+}
+
+function middleware(req, res, next) {
+  const startTime = new Date()
+
+  // methods to allow the setting of additional information to be logged for the request
+  req._logInfo = {}
+  req._logMessage = 'http request'
+  req.addLogField = function(field, value) {
+    req._logInfo[field] = value
+  }
+  req.addLogFields = function(fields) {
+    Object.assign(req._logInfo, fields)
+  }
+  req.setLogMessage = function(message) {
+    req._logMessage = message
+  }
+
+  // override the 'end' method to log and record metrics
+  const end = res.end
+  res.end = function() {
```

Comment:
I am a tiny bit uneasy with monkey patching `res.end`, but equally I can't think of a specific argument against. I wonder if it should be done in a separate middleware, but that might be overkill.

I'd be interested in hearing other people's opinions.

Reply:
Yeah - it's a bit on the ugly side but this is what `metrics-sharelatex` does, so we've been using it for some time. There are two reasons to do it this way:
* It allows us to keep the timing stuff that `metrics-sharelatex` does
* It ensures the logging happens on every request, without having to make sure we call `next()` in every controller method.

AFAIK `express` doesn't otherwise provide any kind of 'the request has now been processed' event.